File: algorithm/tim-sort.js
/**
 * Tim Sort Implementation in JavaScript
 * 
 * Tim sort is a hybrid of merge sort and insertion sort. It's the default
 * sorting algorithm in Python and Java. It finds runs (already sorted sequences),
 * uses insertion sort for small runs, and merges runs together.
 * 
 * Time Complexity: O(n log n) worst case, O(n) best case
 * Space Complexity: O(n)
 */

const MIN_MERGE = 32;

function timSort(arr) {
    const n = arr.length;
    const sorted = [...arr];
    
    // Sort individual subarrays of size MIN_MERGE
    for (let i = 0; i < n; i += MIN_MERGE) {
        insertionSortTim(sorted, i, Math.min(i + MIN_MERGE - 1, n - 1));
    }
    
    // Merge subarrays
    for (let size = MIN_MERGE; size < n; size = 2 * size) {
        for (let left = 0; left < n; left += 2 * size) {
            const mid = left + size - 1;
            const right = Math.min(left + 2 * size - 1, n - 1);
            
            if (mid < right) {
                mergeTim(sorted, left, mid, right);
            }
        }
    }
    
    return sorted;
}

function insertionSortTim(arr, left, right) {
    for (let i = left + 1; i <= right; i++) {
        const key = arr[i];
        let j = i - 1;
        while (j >= left && arr[j] > key) {
            arr[j + 1] = arr[j];
            j--;
        }
        arr[j + 1] = key;
    }
}

function mergeTim(arr, left, mid, right) {
    const leftArr = arr.slice(left, mid + 1);
    const rightArr = arr.slice(mid + 1, right + 1);
    
    let i = 0, j = 0, k = left;
    
    while (i < leftArr.length && j < rightArr.length) {
        if (leftArr[i] <= rightArr[j]) {
            arr[k] = leftArr[i];
            i++;
        } else {
            arr[k] = rightArr[j];
            j++;
        }
        k++;
    }
    
    while (i < leftArr.length) {
        arr[k] = leftArr[i];
        i++;
        k++;
    }
    
    while (j < rightArr.length) {
        arr[k] = rightArr[j];
        j++;
        k++;
    }
}

// Example usage
if (require.main === module) {
    const arr = [64, 34, 25, 12, 22, 11, 90];
    console.log('Tim Sort:', timSort(arr));
}

module.exports = { timSort };

File: algorithm/tim-sort.py
"""
Tim Sort Implementation in Python

Tim sort is a hybrid of merge sort and insertion sort. It's the default
sorting algorithm in Python and Java. It finds runs (already sorted sequences),
uses insertion sort for small runs, and merges runs together.

Time Complexity: O(n log n) worst case, O(n) best case
Space Complexity: O(n)
"""

MIN_MERGE = 32


def tim_sort(arr):
    n = len(arr)
    sorted_arr = arr.copy()
    
    # Sort individual subarrays of size MIN_MERGE
    for i in range(0, n, MIN_MERGE):
        insertion_sort_tim(sorted_arr, i, min(i + MIN_MERGE - 1, n - 1))
    
    # Merge subarrays
    size = MIN_MERGE
    while size < n:
        for left in range(0, n, 2 * size):
            mid = left + size - 1
            right = min(left + 2 * size - 1, n - 1)
            
            if mid < right:
                merge_tim(sorted_arr, left, mid, right)
        
        size *= 2
    
    return sorted_arr


def insertion_sort_tim(arr, left, right):
    for i in range(left + 1, right + 1):
        key = arr[i]
        j = i - 1
        while j >= left and arr[j] > key:
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = key


def merge_tim(arr, left, mid, right):
    left_arr = arr[left:mid + 1]
    right_arr = arr[mid + 1:right + 1]
    
    i = j = 0
    k = left
    
    while i < len(left_arr) and j < len(right_arr):
        if left_arr[i] <= right_arr[j]:
            arr[k] = left_arr[i]
            i += 1
        else:
            arr[k] = right_arr[j]
            j += 1
        k += 1
    
    while i < len(left_arr):
        arr[k] = left_arr[i]
        i += 1
        k += 1
    
    while j < len(right_arr):
        arr[k] = right_arr[j]
        j += 1
        k += 1


if __name__ == "__main__":
    arr = [64, 34, 25, 12, 22, 11, 90]
    print("Tim Sort:", tim_sort(arr))

